subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1406,12 +1390,175 @@ ssize_t ztls_recvfrom(int sock, void *buf, size_t max_len, int flags,
 
 	if (ret == MBEDTLS_ERR_SSL_WANT_READ ||
 	    ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
-		errno = EAGAIN;
+		ret = -EAGAIN;
+	} else {
+		ret = -EIO;
+	}
+
+	errno = -ret;
+	return -1;
+}
+
+#if defined(CONFIG_NET_SOCKETS_ENABLE_DTLS)
+static ssize_t recvfrom_dtls_client(struct net_context *context, void *buf,
+				    size_t max_len, int flags,
+				    struct sockaddr *src_addr,
+				    socklen_t *addrlen)
+{
+	int ret;
+
+	if (!context->tls->tls_established) {
+		ret = -ENOTCONN;
+		goto error;
+	}
+
+	ret = mbedtls_ssl_read(&context->tls->ssl, buf, max_len);
+	if (ret >= 0) {
+		if (src_addr && addrlen) {
+			dtls_peer_address_get(context, src_addr, addrlen);
+		}
+		return ret;
+	}
+
+	switch (ret) {
+	case MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY:
+		ret = -ECONNRESET;
+		break;
+
+	case MBEDTLS_ERR_SSL_TIMEOUT:
+		(void)mbedtls_ssl_close_notify(&context->tls->ssl);
+		ret = -ECONNABORTED;
```

Comment:
A timeout error wouldn't be a better choice?

Reply:
Perhaps, will update.